### PR TITLE
refactor: Add NormalValue

### LINKED ISSUE #2403: Have a single place in code that manages standard (or normal) types
At the moment we have few places in code that deal with values of types like `int64`, `int32`, `int`... and converts it to standard on `int64`.

We should have a single place in code that takes care of converting a value of any type into a standard (used by GraphQL) type. So that the value that is being passed around is known be of standard (or normal) type and there is no need to do lengthy type assertions.

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #2403

## Description

This change introduces `NormalValue` type that carries a standard (or normal) value that doesn't need to be type-asserted with all possible types.

### CODE REVIEW

File: client/normal_value.go
Comment (AndrewSisley) [inline client/normal_value.go:21]: suggestion: It might be worth chatting over discord about this before commiting too much time to polishing it, I have a few concerns that need talking over and others may have some too.  My biggest are below:

1. This means a pointer for every field value
2. Loss of compile time safety

You might actually be able to remove the second concern by using a constrained generic instead of `any`, in most langs this would also resolve (1), but IIRC in Go all generic variables are pointers anyway (I'm not sure, worth checking):
```
type NormalType interface {
  int64 | float64 | string | etc
}

type NormalValue[T NormalType]  struct {
  value T
}
```

Comment (fredcarle) [reply client/normal_value.go:21]: > in Go all generic variables are pointers anyway

I don't think that's true. Do you remember where you might have seen this?

Comment (islamaliev) [reply client/normal_value.go:21]: > It might be worth chatting over discord

I actually wanted to bring it up yesterday, but the time switch caught me off the guard. And then we want over the time.

> Loss of compile time safety

We use `any` everywhere and so far it wasn't a concern.
In this case `NormalValue` is meant to provide some type safety in more compact and reliable way.
It it mostly a wrapper over the value held by `client.FieldValue` which is also of `any` type.

Comment (fredcarle) [reply client/normal_value.go:21]: > It is mostly a wrapper over the value held by client.FieldValue which is also of any type.

Could we not apply all the `NormalValue` methods directly to `FieldValue`?

Comment (islamaliev) [reply client/normal_value.go:21]: > in Go all generic variables are pointers anyway

I might need to do some additional research on this, but I hardly believe.
In any language with decent support of generics (like C++ or Rust) they are all resolved at compile time and every instantiation of a generic type is a distinct type.

Comment (AndrewSisley) [reply client/normal_value.go:21]: https://deepsource.com/blog/go-1-18-generics-implementation

> In Boxing, "values" are boxed and passed around as references of the polymorphic type, generally using a pointer table, commonly called virtual table. This is how Go's interface are implemented. This generally produces a smaller binary and requires short compile-times but may have a toll on the runtime performance.
> 
> Go generics combines concepts from "monomorphisation" (stenciling) and "boxing" (dynamic dispatch) and is implemented using [GCshape stenciling and dictionaries](https://github.com/golang/proposal/blob/master/design/generics-implementation-dictionaries-go1.18.md). This allows Go to have fast compile times and smaller binaries while having generics.

Go generics are not the same generics that you find in Rust/C#/C++/etc - they have a runtime cost here :(

Comment (AndrewSisley) [reply client/normal_value.go:21]: > We use any everywhere and so far it wasn't a concern.

This is true, when I think about it our code is all very reliant on boxing anyway, however I think the proposed code might result in double boxing - e.g. in a `core.Doc.Fields` we have an array of `any`, which here would box `NormalValue`, which in turn boxes the field value.

If we are able to change the type of `core.Doc.Fields` to `NormalValue` we might only have a single box (my preference being the generic, not an any anyway).

File: client/normal_value.go
Comment (AndrewSisley) [inline client/normal_value.go:1138]: thought: I don't think this should exist no?  If a `nil` is required, of an unknown type, would `NormalVoid` not be more appropriate?

File: client/normal_value.go
Comment (AndrewSisley) [inline client/normal_value.go:1146]: praise: This is a really nice func signature :)

File: client/normal_value.go
Comment (AndrewSisley) [inline client/normal_value.go:11]: suggestion: This file is pretty hard to read, and I can see it getting very disorganized over time, and very hard to find stuff/see what exists.  I suggest splitting it up by type in its own directory within client:

``` client/value
void.go
nil.go
int.go
// etc
```
If you really want the types directly in the client package, you could always re-export them(or just the constructors+interface?) Rust style from a mod.rs style file (client/normal_value.go):
```
type NormalValue = value.NormalValue
func NewNormalInt(...) { // I don't know if there is a better way than this in Go for funcs?
  return value.NormalValue()
}
```

File: client/normal_value.go
Comment (AndrewSisley) [inline client/normal_value.go:20]: todo: Please document all the public stuff before merge.  If you end up breaking up this file, a doc.go with package level documentation might be nice too.

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
I'm still reviewing, but submitting a few suggestions now that may aid further review :)

**Review by fredcarle — COMMENTED** [review]
Currently reviewing but I just want to point out that this PR is in conflict with #2414. #2414 should merge first and the conflict can be resolved here.

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:11]: I also thought about it but then it would need access to `client.Document` and `client` package will also depend on it. You have a suggestion?

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:20]: will do

File: client/normal_value.go
Comment (fredcarle) [reply client/normal_value.go:11]:
```
type NormalValue = value.NormalValue  
func NewNormalInt(...) { // I don't know if there is a better way than this in Go for funcs?  
  return value.NormalValue()  
} 
```

Please don't 😅.

I'm fine if we go for a client sub-package. `client/normalized` would be my choise. Then we could have the interface`normalized.Value` and the types `normalized.String`, `normalized.Int`, etc.

Comment (fredcarle) [reply client/normal_value.go:11]: > I also thought about it but then it would need access to client.Document and client package will also depend on it. You have a suggestion?

No need to bring `client.Document` into the sub package. Just have the `normilized.Value` interface implement for `client.Document` where `client.Document` is defined.

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:1138]: NormalVoid is like a dummy returning default value like `false`. It there just to rid other type of necessary to implement all methods.
And I think having a dedicated `NewNormalNil` more explicitly conveys the intent.

File: client/normal_value.go
Comment (fredcarle) [reply client/normal_value.go:11]: Sorry. I need to point out that for the last comment to work we need to keep the interface in client. So the `normalized` package wouldn't have the `normalized.Value` interface. It would just hold the types.

Comment (islamaliev) [reply client/normal_value.go:11]: > No need to bring client.Document into the sub package. Just have the normilized.Value interface implement for client.Document where client.Document is defined.

Then it won't be possible to get normal value from `NewNormalValue` passing in a `*client.Document`. Not sure it's a good trade-off.

File: client/normal_value.go
Comment (AndrewSisley) [reply client/normal_value.go:1138]: > NormalVoid is like a dummy returning default value like false.

Yes, that's what I mean, is why I'd prefer it over `NormalNil` - `NormalNil` looks like it is typed, when it is not really - it also returns true for IsNillable no? (unwanted IMO if the type is undefined).

File: client/normal_value.go
Comment (AndrewSisley) [reply client/normal_value.go:11]: An alternative could be to namespace the file names within the client package: `_normal_int.go`, `_normal_value.go`, etc - please dont do this without a namespace though, or we wont be able to find anything else in the directory.

Comment (fredcarle) [reply client/normal_value.go:11]: That might be the more sensible approach. No need for the `_` prefix though.

Comment (islamaliev) [reply client/normal_value.go:11]: is "normal" bad namespace? `normal_int.go`, `normal_nillable_int.go`...?

Comment (fredcarle) [reply client/normal_value.go:11]: > is "normal" bad namespace? `normal_int.go`, `normal_nillable_int.go`...?

I much prefer just `normal_...` 👍

Comment (fredcarle) [reply client/normal_value.go:11]: > I would prefer that they are all either at the end alphabetically, or at the start - otherwise the other files will be split on either side.

I'm pretty sure the compiler ignores files with `_` prefixes just like it ignores  files with a `.` prefix.

Comment (fredcarle) [reply client/normal_value.go:11]: I confirm https://github.com/golang/go/blob/master/src/go/build/build.go#L548

Comment (AndrewSisley) [reply client/normal_value.go:11]: :(

`zzz_normal_int.go` :laughing: ? I will have to tolerate them being in the middle of the package them :(

File: client/normal_value.go
Comment (fredcarle) [reply client/normal_value.go:1138]: It seems that everywhere this is used, we have access to the field kind which means that we can get the nillable supported type directly instead of a catch all `nil` type. I think all we would need is an extra `NewNormalValueFromKind` function.

Comment (fredcarle) [reply client/normal_value.go:1138]: `DecodeFieldValue` would need to take a FieldKind as parameter as a side effect.

File: tests/integration/index/query_with_index_only_field_order_test.go
Comment (fredcarle) [inline tests/integration/index/query_with_index_only_field_order_test.go:111]: thought: That was a sneeky bug that we should have caught 😅

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
Submitting a handful of documentation requests now, interface looks good :)

File: client/normal_value.go
Comment (fredcarle) [reply client/normal_value.go:32]: > I would not expect it to return `immutable.Option`

It doesn't. Tt returns the value of the `immutable.Option`

Comment (AndrewSisley) [reply client/normal_value.go:32]: Yes... How did I miss read that - maybe, my eyes half read the docs for Any part way through...

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:41]: nothin is converted there because the value is stored as `int64` all conversions happen either in `NewNormalValue` or `NewNormalInt`

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:54]: there is actually a note in the documentation of the interface itself. I thought it would be enough.

File: client/normal_value.go
Comment (AndrewSisley) [reply client/normal_value.go:41]: Yes, but can can you please document that on the interface?

File: client/normal_value.go
Comment (AndrewSisley) [reply client/normal_value.go:54]: Ah true, I know it is a hassle, but can you please document it on the funcs too - I think that is best practice, and it is the first place where people will look most of the time to try and find that out.

Comment (AndrewSisley) [reply client/normal_value.go:54]: I think the time spent now will be much smaller than the sum of all the head scratching (or bugs) that could take place if people have to chase down this info (or guess and potentially get it wrong)

File: client/value.go
Comment (AndrewSisley) [inline client/value.go:65]: thought: This seems very broken and very much looks like a bug to me.  Please let me know if you know it is not a bug.

File: core/encoding.go
Comment (AndrewSisley) [inline core/encoding.go:250]: suggestion: This check seems unnecessary - did something break without it?  How could it ever happen?

Something would have to be very very broken for this to ever be false no?

I suggest removal of the `if`.

todo: The comment above has been moved away from the line upon which it is commenting, please move it back :)

File: core/encoding.go
Comment (AndrewSisley) [inline core/encoding.go:266]: todo: Please document how/when this can happen, or remove it.  I am struggling to imagine how this might happen (besides manually messing around in the datastore).

File: core/key_test.go
Comment (AndrewSisley) [inline core/key_test.go:335]: todo: Please remove or uncomment the commented out tests.

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:26]: ok

File: client/normal_value.go
Comment (AndrewSisley) [inline client/normal_value.go:70]: suggestion: `IsArray` might be slightly nicer to have futher up this definition, next to `IsNillable`.

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [inline db/fetcher/indexer_iterators.go:292]: question: This looks like a potential behaviour change to me, we are not checking for `nil` here - is that alright here?

suggestion: Maybe a func like the below would be handy (annoyingly generic funcs cant be defined on Go interfaces or we could replace most funcs on it with a couple of generic ones...):
```
func AsT[T client.NormalValueConstraints](v NormalValue) (T, bool) {
  any := v.Unwrap()
  return any.(T)
}
```
It could either live here as a private func, or in the client package, then in the funcs here you wouldn't need two ifs, you could just:
```
func (m *intMatcher) Match(value client.NormalValue) (bool, error) 
  intVal, ok := value.AsT[int64]()
  if ok {
    return m.evalFunc(intVal, m.value), nil
  }
  return false, NewErrUnexpectedTypeValue[int64](value)
}
```

File: client/normal_nil.go
Comment (AndrewSisley) [inline client/normal_nil.go:46]: suggestion: Maybe returning `NormalVoid` would be nicer than an error here? I'm fairly torn though, so don't do this if you are even slightly preferring the error :)

File: encoding/field_value.go
Comment (AndrewSisley) [inline encoding/field_value.go:84]: question: Why did you remove the Kind from the log, it is still known here and might be useful no?

File: client/normal_value.go
Comment (AndrewSisley) [inline client/normal_value.go:27]: suggestion: A `Kind() FieldKind` func might be handy to have on this.

**Review by AndrewSisley — COMMENTED** [review]
Looks good :)  Just a couple of small suggestions :)

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:54]: > document whether they return nillable values for none nillable `NormalValues`

I don't think we should document on what it returns in case it's nillable or not nillable. It's irrelevant. If the requested method does not match exactly the underlying type, the second param will be false (even if both underlying and requested are nillable). And in this case the first parameter is the default value as with standard Go type assertion. We don't document that the first value of `Int()` returns 0, because it should be clear and nobody is supposed to use this value anyway.

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:72]: as I mentioned earlier, I don't think it's worth documenting on the first value if the second returns false.
It acts as Go's type assertion that in this case returns default values (or zero). For arrays it's nil

File: client/normal_value.go
Comment (AndrewSisley) [reply client/normal_value.go:54]: > I don't think we should document on what it returns in case it's nillable or not nillable. It's irrelevant. If the requested method does not match exactly the underlying type, the second param will be false (even if both underlying and requested are nillable).

This is not clear from the function signature.  It is not unreasonable to expect `NillableInt` to return Some when calling it on a none nillable int type.

File: client/normal_void.go
Comment (islamaliev) [reply client/normal_void.go:19]: no, it's not used for anything else. And it doesn't implement `Unwrap` method (former `Any()`) to enforce override if `NormalVoid` is ever embedded by a custom type.

File: client/value.go
Comment (islamaliev) [reply client/value.go:65]: where exactly is do you think there can be a bug? switching from go's type assertion to `NormalValue` type assertion? They are identical.

Comment (AndrewSisley) [reply client/value.go:65]: There is very likely a bug in develop branch.

File: db/fetcher/indexer_iterators.go
Comment (fredcarle) [reply db/fetcher/indexer_iterators.go:292]: > annoyingly generic funcs cant be defined on Go interfaces or we could replace most funcs on it with a couple of generic ones...

They can be defined on interfaces: https://go.dev/play/p/gt5p6hQsq_Q

Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go:292]: > They can be defined on interfaces:

Only if the type param is declared on the interface, which you dont want to do here.

Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:292]: I don't a benefit here. Why can't you just write `v, ok := value.Unwrap().(int64)`?

Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go:292]: > Why can't you just write v, ok := value.Unwrap().(int64)?

That is my suggestion, only wrapped up into a func so you don't have to cast in a whole bunch of places.

File: encoding/field_value.go
Comment (islamaliev) [reply encoding/field_value.go:84]: I removed it because at some point `cliend.FieldKind` was removed from the func signature.
I bought it back.

File: client/value.go
Comment (AndrewSisley) [reply client/value.go:65]: I just looked it over - I misread, it (`convertImmutable`) is weirdly written (probably by me...) and does append the nil values instead of setting by index.  Resolving this conversation.

File: core/key_test.go
Comment (islamaliev) [reply core/key_test.go:335]: ok

File: core/encoding.go
Comment (islamaliev) [reply core/encoding.go:250]: no, nothing breaks. The two slices should have the same len.
I just thought some abnormal scenario (that I can't think of) these might be different.
But I think you are right, it is not necessary. I bought it back.

File: core/encoding.go
Comment (islamaliev) [reply core/encoding.go:266]: removed

File: client/normal_nil.go
Comment (islamaliev) [reply client/normal_nil.go:46]: I'm not sure about it at the moment. I think we don't have enough use cases to see if it makes sense. I'd say lets stay on the safe side and keep the error for now.

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:27]: good suggestion but we don't have all `FieldKind` to cover all possible normal values. For example, we don't have `FieldKind_INT`.

File: client/normal_new.go
Comment (fredcarle) [inline client/normal_new.go:21]: todo: "`if` the given type is..."

File: client/normal_new.go
Comment (fredcarle) [inline client/normal_new.go:22]: todo: "`it` with `go` through..."

File: client/normal_new.go
Comment (fredcarle) [inline client/normal_new.go:23]: question: Is `minmal` the right word here? There might be something missing to make it make sense.

File: client/normal_new.go
Comment (fredcarle) [inline client/normal_new.go:31]: thought: If we passed `FieldKind` in here it would make it even more strongly typed given that `[]any{int32(1), int64(2)}` could very well be `[]immutable.Option[int64]{immutable.Some(1), immutable.Some(2)}` instead of `[]int64{1, 2}`. This can be done in a separate PR though.

File: client/normal_util.go
Comment (fredcarle) [inline client/normal_util.go:13]: todo: "`if` the given value..."

File: client/normal_value.go
Comment (fredcarle) [inline client/normal_value.go:24]: todo: "to Go's type assertion" or "to the Go type assertion"

File: db/fetcher/indexer_iterators.go
Comment (fredcarle) [reply db/fetcher/indexer_iterators.go:292]: `v, ok := value.Unwrap().(int64)` this wouldn't work as it wouldn't check for `nil` passing `0` to the `evalFunc` is no a correct behaviour if the actual value is `nil`.

**Review by fredcarle — APPROVED** [review]
Approving and assuming you'll resolve all comments before merge :)

This is overall really nice and brings us to a place where we can easily add on if we need to as we go.

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go:292]: Ah true, the private func would need to be:
```
func AsT[T client.NormalValueConstraints](v NormalValue) (T, bool) {
  if v.IsNil() {
    return false
  }
  return v.Unwrap().(T), true
}
```

File: client/normal_value.go
Comment (AndrewSisley) [reply client/normal_value.go:27]: I think that can be sorted out when we add support for `FieldKind_INT`

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:56]: done

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:78]: done

File: client/normal_value.go
Comment (islamaliev) [reply client/normal_value.go:27]: this is somewhat involved addition. I've created another task for it https://github.com/sourcenetwork/defradb/issues/2463

File: client/normal_new.go
Comment (islamaliev) [reply client/normal_new.go:31]: this is somewhat involved addition. I've created another task for it https://github.com/sourcenetwork/defradb/issues/2463

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:293]: I just changed code to check it explicitly. You suggestion wouldn't work as it will return 0 for `nil` values which will match if the user runs a filter `_eq: 0`

**Pull request merged**
